### PR TITLE
Stop double term sanative

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket:  https://core.trac.wordpress.org/ticket/58329
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-term.php [hunk 1]
```diff
@@ -109,11 +109,12 @@ final class WP_Term {
 	 * @param int    $term_id  Term ID.
 	 * @param string $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
 	 *                         disambiguating potentially shared terms.
+	 * @param string $filter   Optional. How to sanitize term fields. Default 'raw'.
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php]: The `$filter` parameter also supports `false` which should be included here.

File: src/wp-includes/class-wp-term.php [hunk 2]
```diff
@@ -109,11 +109,12 @@ final class WP_Term {
...
 	 * @return WP_Term|WP_Error|false Term object, if found. WP_Error if `$term_id` is shared between taxonomies and
 	 *                                there's insufficient data to distinguish which term is intended.
 	 *                                False for other failures.
 	 */
-	public static function get_instance( $term_id, $taxonomy = null ) {
+	public static function get_instance( $term_id, $taxonomy = null, $filter = 'raw' ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php:118]:
```
* @since 6.4.0 Added the `$filter` parameter.
```

Add missing since for new parameter.

File: tests/phpunit/tests/term/wpTerm.php [hunk 3]
```diff
@@ -5,6 +5,17 @@
  */
 class Tests_Term_WpTerm extends WP_UnitTestCase {
 	protected static $term_id;
+	/**
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term/wpTerm.php:9]:
```suggestion
	
	/**
```
Add line break

**Review by mukeshpanchal27 — CHANGES_REQUESTED** [review]
Thanks @spacedmonkey for the PR, I left some nitpick. Major issue it `get_term` filter is not working after the new code changes.

**Comment by spacedmonkey** [comment]
Can you define not working? Steps to replicate!

**Comment by mukeshpanchal27** [comment]
@spacedmonkey Could you please check the unit test that are failing.

File: src/wp-includes/class-wp-term.php [hunk 4]
```diff
@@ -103,17 +103,19 @@ final class WP_Term {
...
 	 *
 	 * @since 4.4.0
+	 * @since 6.4.0 Added the `$filter` parameter.
 	 *
 	 * @global wpdb $wpdb WordPress database abstraction object.
 	 *
 	 * @param int    $term_id  Term ID.
 	 * @param string $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
 	 *                         disambiguating potentially shared terms.
+	 * @param string|false $filter   Optional. How to sanitize term fields. Default 'raw'.
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php]:
```suggestion
	 * @param int          $term_id  Term ID.
	 * @param string       $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
	 *                               disambiguating potentially shared terms.
	 * @param string|false $filter Optional. How to sanitize term fields. Default 'raw'.
```

Correct indention for `param`

File: src/wp-includes/class-wp-term.php [hunk 5]
```diff
@@ -195,7 +198,14 @@ public static function get_instance( $term_id, $taxonomy = null ) {
 	 * @param WP_Term|object $term Term object.
 	 */
 	public function __construct( $term ) {
+		$int_fields = array( 'parent', 'term_id', 'count', 'term_group', 'term_taxonomy_id', 'object_id' );
 		foreach ( get_object_vars( $term ) as $key => $value ) {
+			if ( in_array( $key, $int_fields, true ) ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php:203]:
```suggestion
			if ( isset( $int_fields[ $key ] ) ) {
```
Use `isset` instead `in_array`

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey for the update. Left some additional feedbacks.

File: src/wp-includes/class-wp-term.php [hunk 6]
```diff
@@ -182,8 +184,9 @@ public static function get_instance( $term_id, $taxonomy = null ) {
 		}
 
 		$term_obj = new WP_Term( $_term );
-		$term_obj->filter( $term_obj->filter );
-
+		if ( $filter ) {
+			$term_obj->filter( $filter );
+		}
```
Comment (costdev) [inline src/wp-includes/class-wp-term.php]: The previous argument passed to `::filter()` was `$term_obj->filter`, and the new one is `$filter`.

Given a `WP_Term` object with its `$filter` property set to `'rss'`:
```php
$my_term = new WP_Term(
    (object) array(
        'filter' => 'rss',
        // More data.
    )
); // Assume term_id = 1
```
And this call to `get_term()`:

```php
$term = get_term( 1 );
```
Previously, this would have been filtered for `'rss'` in `WP_Term::get_instance()`, then filtered for `'raw'` in `get_term()` before being returned. Now, `WP_Term::get_instance()` won't filter it for `'rss'` before returning to `get_term()`. This could lead to a different return value.

I think we'll need to make sure there's PHPUnit tests covering this scenario and similar ones to make sure we're not implementing a BC break here.

Can pass`'filter' => 'rss'` when creating the term, then hook `term_{$field}_rss` to modify the value, and finally test that calling `get_term()` returns the same value for the given field before/after the patch.

Similarly, using the same term, `get_term()` should be called using `'db'` for the `$filter` parameter, so that neither `get_term()` nor `WP_Term::get_instance()` are using their defaults, followed by testing the field's value before/after the patch.

File: src/wp-includes/class-wp-term.php (hunk 5 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-term.php:203]: This doesn't work. Isset is used for checking if a key existings. `$int_fields` is a numberic array, this does not work.

File: src/wp-includes/class-wp-term.php (hunk 6 above)
Comment (spacedmonkey) [reply src/wp-includes/class-wp-term.php]: @costdev Could you provide a code example of the issue you documented? I am finding it hard to replicate the above.

So called `get_term` as a parameter for filter, that would be respected. `WP_Term::get_instance` stores term objects in object cache in the raw filter. 

The only issue related here would doing this. 

```
$my_term = new WP_Term(
    array(
        'filter' => 'rss',
    )
);
$my_term::get_instance(123);
```

This would no longer apply the rss filter. The above feels a little edge case, but I may look into it.

Comment (costdev) [reply src/wp-includes/class-wp-term.php]: Sorry for the delayed response! `WP_Term::$filter` not being respected was the issue I was mentioning.

This can be demonstrated by running this test with/without the patch:
```php
public function test_get_term_should_respect_term_filter_property() {
	add_filter(
		'term_name_rss',
		static function() {
			return 'RSS Name';
		}
	);

	$term = new WP_Term(
		(object) array(
			'term_id' => 999,
			'name'    => 'My Test Term Name',
			'filter'  => 'rss',
		)
	);

	// Save to cache for later retrieval.
	wp_cache_set( $term->term_id, $term, 'terms' );

	// Test the name before running `get_term()`.
	$this->assertSame( 'My Test Term Name', $term->name, 'The original term name was not set.' );

	// Retrieve later.
	$actual = get_term( $term->term_id );

	// Test that the filter property has been respected.
	$this->assertSame( 'RSS Name', $actual->name, 'The "rss" filter was not applied despite being the term filter.' );
}
```
The above test is one way to reproduce the issue. There may be other ways in which a term's `$filter` property is set before the term is filtered.

Some notes that might be pertinent when investigating this issue:
- The `$filter` property has `public` visibility.
- `sanitize_term()` will change the `$filter` property to match `$context` value.
- `'rss'` is used just as an example of a non-`'raw'` filter value. The other possible values are:
  - `'edit'`
  - `'db'`
  - `'display'`
  - `'attribute'`
  - `'js'`

Comment (spacedmonkey) [reply src/wp-includes/class-wp-term.php]: @costdev That test is invalid. 

Look at the following. 

https://github.com/WordPress/wordpress-develop/blob/da28362ba1a99a86211dfc75c376b16f8218b967/src/wp-includes/class-wp-term.php#L176-L181

As you can see the term object stored as filter = raw. There is never a case where it would be stored filtered for rss.

File: src/wp-includes/class-wp-term.php [hunk 7]
```diff
@@ -103,17 +103,19 @@ final class WP_Term {
...
 	 *
 	 * @global wpdb $wpdb WordPress database abstraction object.
 	 *
-	 * @param int    $term_id  Term ID.
-	 * @param string $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
-	 *                         disambiguating potentially shared terms.
+	 * @param int          $term_id  Term ID.
+	 * @param string       $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
+	 *                               disambiguating potentially shared terms.
+	 * @param string|false $filter Optional. How to sanitize term fields. Default 'raw'.
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php]:
```suggestion
	 * @param string       $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
	 *                               disambiguating potentially shared terms.
	 * @param string|false $filter   Optional. How to sanitize term fields. Default 'raw'.
```

Document alignment.

File: src/wp-includes/class-wp-term.php [hunk 8]
```diff
@@ -195,7 +198,14 @@ public static function get_instance( $term_id, $taxonomy = null ) {
 	 * @param WP_Term|object $term Term object.
 	 */
 	public function __construct( $term ) {
+		$int_fields = array( 'parent', 'term_id', 'count', 'term_group', 'term_taxonomy_id', 'object_id' );
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php:201]: Could you please add notes for these changes so that others can understand why these changes are needed?

```
@since 6.4.0 Notes for changes
```

File: tests/phpunit/tests/term/wpTerm.php [hunk 9]
```diff
@@ -89,4 +101,36 @@ public function test_get_instance_should_respect_taxonomy_when_term_id_is_found_
 		$found = WP_Term::get_instance( self::$term_id, 'wptests_tax2' );
 		$this->assertFalse( $found );
 	}
+
+	/**
+	 * @ticket 58329
+	 */
+	public function test_get_instance_no_sanitize() {
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/term/wpTerm.php]: Could we consider adding a test with a data provider? In the data provider, I've included a test case, and if it's not necessary, we can remove it. Thank you.

```
	/**
	 * @ticket 58329
	 *
	 * @dataProvider data_get_instance_with_sanitize
	 *
	 * @param string|false $sanitize Sanitize term fields.
	 */
	public function test_get_instance_with_sanitize( $sanitize ) {
		$obj           = new stdClass();
		$obj->taxonomy = 'wptests_tax';
		foreach ( self::$int_fields as $int_field ) {
			$obj->$int_field = 'test';
		}
		wp_cache_add( 999, $obj, 'terms' );
		$term = WP_Term::get_instance( 999, 'wptests_tax', $sanitize );
		foreach ( self::$int_fields as $int_field ) {
			$this->assertIsInt( $term->$int_field );
		}
	}

	/**
	 * Data provider.
	 *
	 * @return array[]
	 */
	public function data_get_instance_with_sanitize() {
		return array(
			'check false value for sanitize'      => array( false ),
			'check empty value for sanitize'      => array( '' ),
			'check raw value for sanitize'        => array( 'raw' ),
			'check any string value for sanitize' => array( 'test' ),
		);
	}
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thank you, @spacedmonkey, for making the changes. They look solid to me. I have some more feedback to provide, and once those adjustments are made, it should be ready for approval from my side.

File: src/wp-includes/class-wp-term.php [hunk 10]
```diff
@@ -182,8 +184,9 @@ public static function get_instance( $term_id, $taxonomy = null ) {
 		}
 
 		$term_obj = new WP_Term( $_term );
-		$term_obj->filter( $term_obj->filter );
-
+		if ( $filter ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-term.php]: Do we need to check if `$filter` should be `row` then it will allow for sanitize?

```suggestion
		if ( 'raw' === $filter ) {
```

File: src/wp-includes/taxonomy.php [hunk 11]
```diff
@@ -944,7 +944,7 @@ function get_term( $term, $taxonomy = '', $output = OBJECT, $filter = 'raw' ) {
 			$_term = sanitize_term( $term, $taxonomy, 'raw' );
 			$_term = new WP_Term( $_term );
 		} else {
-			$_term = WP_Term::get_instance( $term->term_id );
+			$_term = WP_Term::get_instance( $term->term_id, $taxonomy );
```
Comment (spacedmonkey) [inline src/wp-includes/taxonomy.php]:
```suggestion
			$_term = WP_Term::get_instance( $term->term_id );
```

File: tests/phpunit/tests/term/wpTerm.php [hunk 12]
```diff
@@ -89,4 +101,39 @@ public function test_get_instance_should_respect_taxonomy_when_term_id_is_found_
...
+	 * @param string|false $sanitize Sanitize term fields.
+	 */
+	public function test_get_instance_with_sanitize( $sanitize ) {
+		$obj           = new stdClass();
+		$obj->taxonomy = 'wptests_tax';
+		foreach ( self::$int_fields as $int_field ) {
+			$obj->$int_field = 'test';
+		}
+		wp_cache_add( 999, $obj, 'terms' );
+		$term = WP_Term::get_instance( 999, 'wptests_tax', $sanitize );
```
Comment (joemcgill) [inline tests/phpunit/tests/term/wpTerm.php:119]: I don't understand what this code is trying to test. I assumed that this new parameter that you're adding should map to one of the filter contexts described in the [WP_Term::filter](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-includes/class-wp-term.php#L203-L212) method, but the values being passed by the data provider seem to be something different. Can you help clarify?

File: src/wp-includes/class-wp-term.php (hunk 8 above)
Comment (joemcgill) [reply src/wp-includes/class-wp-term.php:201]: I agree with @mukeshpanchal27. I'm not clear on why some early sanitization is being applied in the constructor. In fact, this seems to duplicate logic that is already being applied during [term filed sanitization](https://github.com/WordPress/wordpress-develop/blob/6.3.1/src/wp-includes/taxonomy.php#L1708-L1714), so this seems unnecessary.

File: tests/phpunit/tests/term/wpTerm.php [hunk 13]
```diff
@@ -89,4 +101,39 @@ public function test_get_instance_should_respect_taxonomy_when_term_id_is_found_
...
+	 *
+	 * @param string|false $sanitize Sanitize term fields.
+	 */
+	public function test_get_instance_with_sanitize( $sanitize ) {
+		$obj           = new stdClass();
+		$obj->taxonomy = 'wptests_tax';
+		foreach ( self::$int_fields as $int_field ) {
+			$obj->$int_field = 'test';
+		}
+		wp_cache_add( 999, $obj, 'terms' );
```
Comment (joemcgill) [inline tests/phpunit/tests/term/wpTerm.php:118]: Nitpick, but rather than setting up your test case by adding the Term object to cache, it would be more robust to use the factory to generate a new term, or use the one already generated in the `wpSetUpBeforeClass` method and saved to `self::$term_id` (as long as that doesn't introduce side effects to other test cases).

File: src/wp-includes/class-wp-term.php [hunk 14]
```diff
@@ -103,17 +103,19 @@ final class WP_Term {
...
+	 * @param int          $term_id  Term ID.
+	 * @param string       $taxonomy Optional. Limit matched terms to those matching `$taxonomy`. Only used for
+	 *                               disambiguating potentially shared terms.
+	 * @param string       $filter Optional. How to sanitize term fields. Default 'raw'.
 	 * @return WP_Term|WP_Error|false Term object, if found. WP_Error if `$term_id` is shared between taxonomies and
 	 *                                there's insufficient data to distinguish which term is intended.
 	 *                                False for other failures.
 	 */
-	public static function get_instance( $term_id, $taxonomy = null ) {
+	public static function get_instance( $term_id, $taxonomy = null, $filter = 'raw' ) {
```
Comment (joemcgill) [inline src/wp-includes/class-wp-term.php:118]: I'm a bit hesitant about making `raw` the default filter that is applied to existing code calling the `WP_Term::get_instance()` method. This could potentially lead to situations where existing code calling this method fails to have any filter applied, where previously it did (an edge case, for sure). What if instead, the default is null and we update the places in `get_term()` that calls this method to intentionally pass 'raw' instead, knowing that the correct filter value will be applied in `get_term()` after the instance is created?

**Review by joemcgill — COMMENTED** [review]
@spacedmonkey I've left some questions about the approach where I feel like I don't have enough clarity. Also, do you have any data on the relative performance benefit this change is expected to have? I can see from my last 6.3 profiling analysis that `sanitize_term_field` is getting called a lot in our classic theme use cases ([link to spreadsheet](https://docs.google.com/spreadsheets/d/1WICTFAn2pOeVlYgpBzLU8Uh-77zQF3_ZO3N97LLlVdU/edit#gid=802202677)), but am not sure how much this change will reduce that number.

File: src/wp-includes/class-wp-term.php (hunk 8 above)
Comment (oglekler) [reply src/wp-includes/class-wp-term.php:201]: I see the reason for sanitation due to possibility that data stored in cache can be modified by something else. But if this is the case, sanitization needs to be left where it is.

**Comment by spacedmonkey** [comment]
This PR has become overly complex and hard to follow. I don't think it cleanly sorted the problem at hand. 

For that reason, I have started afresh with https://github.com/WordPress/wordpress-develop/pull/5260. I closing this PR out. 
@mukeshpanchal27 @joemcgill @OlaIola

**Comment by spacedmonkey** [comment]
Commited https://github.com/WordPress/wordpress-develop/commit/c1247f6a278bd6a5fec799964177743b5d26a2aa
